lib/lists.nix

```
@@ -86,15 +86,67 @@ rec {
         else op (foldl' (n - 1)) (elemAt list n);
     in foldl' (length list - 1);
 
-  /* Strict version of `foldl`.
+  /*
+    Reduce a list by applying a binary operator from left to right,
+    starting with an initial accumulator.
+    Before each application of the operator, the accumulator value is evaluated.
+    This behavior makes this function stricter than [`foldl`](#function-library-lib.lists.foldl).
 
-     The difference is that evaluation is forced upon access. Usually used
-     with small whole results (in contrast with lazily-generated list or large
-     lists where only a part is consumed.)
+    This strictness also allows avoiding an accumulation of thunks,
+    which would otherwise cause a stack overflow for large enough lists.
 
-     Type: foldl' :: (b -> a -> b) -> b -> [a] -> b
-  */
-  foldl' = builtins.foldl' or foldl;
+    Unlike [`builtins.foldl'`](https://nixos.org/manual/nix/unstable/language/builtins.html#builtins-foldl'),
+    the initial accumulator argument is evaluated before the first iteration.
+
+    A call like
+
+    ```nix
+    foldl' op acc₀ [ x₀ x₁ x₂ ... xₙ₋₁ xₙ ]
+    ```
+
+    is semantically the same as
+
+    ```nix
+    let
+      acc₁   = builtins.seq acc₀   (op acc₀   x₀  );
+      acc₂   = builtins.seq acc₁   (op acc₁   x₁  );
+      acc₃   = builtins.seq acc₂   (op acc₂   x₂  );
+      ...
+      accₙ   = builtins.seq accₙ₋₁ (op accₙ₋₁ xₙ₋₁);
+      accₙ₊₁ = builtins.seq accₙ   (op accₙ   xₙ  );
+    in
+    accₙ₊₁
+
+    # Or ignoring builtins.seq
+    op (op (... (op (op (op acc₀ x₀) x₁) x₂) ...) xₙ₋₁) xₙ
+    ```
+
+    The only difference is that the second version would always cause a stack overflow for large enough lists,
+    which isn't a problem with this function.
```

Stack overflows aren't part of the Nix language semantics, so I'd phrase this a bit differently.
Also I would refrain from comparing these two models, precisely because they're different kinds of models. They are not to be treated as example code.
Furthermore, even if you did, the example with the `let`s and `seq`s will still overflow the stack.

```suggestion
    The most significant benefit of `foldl'` over any hand-written recursion is that it gives the Nix evaluator the opportunity to avoid a stack overflow, even if regular calls are susceptible to this error condition.
```

---

I now removed any text after the code. It's really not that important to point out the difference, nobody would ever write that code anyways.

And the think about stack overflows is already mentioned in an earlier paragraph, so I don't think it needs to be repeated. I did reword that slightly though.